### PR TITLE
8329135: Store Universe::*exception_instance() in CDS archive

### PR DESCRIPTION
**Pull request by iklam**
This is a preparation step for the AOT compiler in the Leyden project. The exception options are frequently accessed by compiled code. Having them inside the CDS archive will allow the AOT to generate smaller/better code.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8329135](https://bugs.openjdk.org/browse/JDK-8329135): Store Universe::*exception_instance() in CDS archive (**Enhancement** - P4)

### Reviewers
 * [Vladimir Ivanov](https://openjdk.org/census#vlivanov) (@iwanowww - **Reviewer**) ⚠️ Review applies to [38f4d732](https://git.openjdk.org/jdk/pull/18551/files/38f4d732e0138fe20832c8fe3b17e877286987ff)
 * [Calvin Cheung](https://openjdk.org/census#ccheung) (@calvinccheung - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/18551/head:pull/18551` \
`$ git checkout pull/18551`

Update a local copy of the PR: \
`$ git checkout pull/18551` \
`$ git pull https://git.openjdk.org/jdk.git pull/18551/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 18551`

View PR using the GUI difftool: \
`$ git pr show -t 18551`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/18551.diff">https://git.openjdk.org/jdk/pull/18551.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/18551#issuecomment-2027486464)

### CODE REVIEW

File: src/hotspot/share/memory/universe.cpp [hunk 1]
```diff
@@ -168,6 +164,52 @@ OopStorage*     Universe::_vm_global = nullptr;
 
 CollectedHeap*  Universe::_collectedHeap = nullptr;
 
+// These are the exceptions that are always created and are guatanteed to exist.
```
Comment (iwanowww) [inline src/hotspot/share/memory/universe.cpp:167]: s/guatanteed/guaranteed/

**Review by iwanowww — APPROVED** [review]
Looks good.

File: src/hotspot/share/memory/universe.cpp [hunk 2]
```diff
@@ -168,6 +164,52 @@ OopStorage*     Universe::_vm_global = nullptr;
...
+// If possible, they can be stored as CDS archived objects to speed up AOT code.
+class BuiltinException {
+  OopHandle _instance;
+  CDS_JAVA_HEAP_ONLY(int _archived_root_index;)
+
+public:
+  BuiltinException() : _instance{} {
```
Comment (calvinccheung) [inline src/hotspot/share/memory/universe.cpp:174]: Is it clearer to do the following?

```
  BuiltinException() {
    _instance();
```

**Review by calvinccheung — APPROVED** [review]

**Comment by iklam** [comment]
Thanks @iwanowww and @calvinccheung for the review
/integrate

**Pull request closed without merging**